### DISCUSSION TITLE
How to listen to all changes to a store?

### DISCUSSION THREAD

**Discussion by Daniel-Khodabakhsh** [solidjs/solid#2058] [Q&A] [answered]
Consider a use-case where a store needs to be processed as a whole whenever any change is made to it. To make it support batching, my idea was to use a `createEffect`. The initial firing should also be avoided so I'm using `createEffect(on(..., { defer: true }))`.

I'm able to achieve this by listening to `JSON.stringify(store)` because stringify deeply interacts with `store`.

```ts
createEffect(on(
  () => JSON.stringify(store),
  (s) => console.log(`JSON.stringify(store): ${s}`),
  { defer: true }
));
```

Here's a [playground I setup](https://playground.solidjs.com/anonymous/e98d8f79-f4b6-4546-9846-d9b5429a854c) with a few `createEffects` under the `store.tsx` file.

Is there a more succinct way of doing this? Using `JSON.stringify` feels a bit too hacky.

**Comment by atk** [comment, marked as answer]
Simply wrap the setter. If that's not possible, use `@solid-primitives/deep`.

**Reply by Daniel-Khodabakhsh** [reply]
Thanks!

Unfortunately wrapping the setters in this case is not a good idea because it is not batch friendly: there would be multiple processing calls (you're suggesting wrapping the setters such that they trigger the processing right?)

Looks like [`@solid-primitives/deep`](https://primitives.solidjs.community/package/deep) has exactly what I need though! Thank you 👍

**Reply by atk** [reply]
You can use a void signal in the wrapper and debounce or throttle it in order to batch updates manually. Have a look at the scheduled package of solid-primitives.

**Reply by dominictobias** [reply]
Sorry what do you mean by "wrap the setter"? Can you show a code example please? Would be great to have some docs on persisting stores, it's a very common scenario.

I'm actually not sure if I can have nested effects but tried outside in a file on one store, also like this:

```ts
// On one store
createRoot(() => {
  createEffect(() => {
    const state = userStore
    console.log('Store state changed:', state)
    const debounce = setTimeout(() => {
      localStorage.setItem(STORE_NAME, JSON.stringify(state))
    }, 1000)
    return () => {
      clearTimeout(debounce)
    }
  })
})

// Tried as component too
export function PersistStores(props: PersistStoresProps) {
  props.stores.forEach(([storeName, store]) => {
    createEffect(() => {
      const storeState = store // Thought accessing here might create an observable
      console.log('Store state changed:', storeName, storeState)
      const debounce = setTimeout(() => {
        localStorage.setItem(storeName, JSON.stringify(storeState))
      }, 1000)
      return () => {
        clearTimeout(debounce)
      }
    })
  })

  return props.children
}
```

But not called on store changes

**Reply by dominictobias** [reply]
Actually `import { trackStore } from '@solid-primitives/deep'` worked, though still curious what I was doing wrong in the vanilla way

**Reply by Daniel-Khodabakhsh** [reply]
@dominictobias what atk meant by wrapping the setter is to create another function that is supposed to be triggered when the store changes, and then calling that whenever you modify the store:

```javascript
function wrappedStoreSetter() {
  setStore(...); // Change the store through its setter

  // Code that reacts to the store change:
  console.log("store changed!");
}
```

and then the last step is to add some debouncing to that call, probably moving that reaction code to another function and debouncing the function call. I didn't get to that because in the end I needed a serialised version of the store so I stuck with `JSON.stringify`
